lib/iris/fileformats/netcdf.py

```diff
@@ -1924,36 +1937,55 @@ def set_packing_ncattrs(cfvar):
             self._dataset.file_format in ('NETCDF3_CLASSIC',
                                           'NETCDF3_64BIT')):
 
-            if packing is None:
-                # Determine whether there is a cube MDI value.
-                fill_value = get_fill_value(cube.core_data())
-
             # Get the values in a form which is valid for the file format.
             data = self._ensure_valid_dtype(cube.data, 'cube', cube)
 
-            if packing is None:
-                dtype = data.dtype.newbyteorder('=')
-
-            # Create the cube CF-netCDF data variable with data payload.
-            cf_var = self._dataset.createVariable(
-                cf_name, dtype, dimension_names,
-                fill_value=fill_value, **kwargs)
-            set_packing_ncattrs(cf_var)
-            cf_var[:] = data
-
+            def store(data, cf_var, fill_value):
+                cf_var[:] = data
+                is_masked = ma.isMaskedArray(data)
+                contains_value = fill_value is not None and fill_value in data
+                return is_masked, contains_value
         else:
-            # Create the cube CF-netCDF data variable.
-            if packing is None:
-                fill_value = get_fill_value(cube.core_data())
-                dtype = cube.dtype.newbyteorder('=')
-
-            cf_var = self._dataset.createVariable(
-                cf_name, dtype,
-                dimension_names, fill_value=fill_value,
-                **kwargs)
-            set_packing_ncattrs(cf_var)
-
-            da.store([cube.lazy_data()], [cf_var])
+            data = cube.lazy_data()
+
+            def store(data, cf_var, fill_value):
+                # Store lazy data and check whether it is masked and contains
+                # the fill value
+                target = _FillValueMaskCheckAndStoreTarget(cf_var, fill_value)
+                da.store([data], [target])
+                return target.is_masked, target.contains_value
+
+        if not packing:
+            dtype = data.dtype.newbyteorder('=')
+
+        # Create the cube CF-netCDF data variable with data payload.
+        cf_var = self._dataset.createVariable(cf_name, dtype, dimension_names,
+                                              fill_value=fill_value,
+                                              **kwargs)
+        set_packing_ncattrs(cf_var)
+
+        # If packing attributes are specified, don't bother checking whether
+        # the fill value is in the data.
+        fill_value_to_check = None if packing else \
```

Comment:
Let's unpack this line. Neat as it is, it really isn't very readable 😉

Reply:
OK